### PR TITLE
DAOS-14348 client: remove dcache entry when a dir is deleted

### PR DESCRIPTION
**Pull request by wiliamhuang**
Features: pil4dfs

Dcache can expire to achieve better consistency. Add a unit test.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: docs/user/filesystem.md [hunk 1]
```diff
@@ -974,6 +974,13 @@ libpil4dfs intercepting summary for ops on DFS:
 [op_sum ]  5003
 ```
 
+### Change dir cache timeout with env
+
+Directory caching is employed for better performance. The default timeout is 10 seconds. User can change it if necessary. The unit is second and an integer should be provided.
+```
+$ export D_IL_DCACHE_TIMEOUT=5
```
Comment (ashleypittman) [inline docs/user/filesystem.md:981]: This is the equivalent setting as dfuse-dentry-dir-time so it would be good to use a matching name.  Perhaps PIL4DFS_DENTRY_DIR_TIME given as this is a env var rather than a container attribute.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 2]
```diff
@@ -5565,6 +5602,12 @@ init_myhook(void)
 		eq_count_max = MAX_EQ;
 	}
 
+	rc = d_getenv_uint64_t("D_IL_DCACHE_TIMEOUT", &dcache_timeout_loc);
+	if (rc != -DER_NONEXIST)
+		dcache_timeout = (uint16_t)dcache_timeout_loc;
+	else
+		dcache_timeout = DCACHE_TIMEOUT;
```
Comment (ashleypittman) [inline src/client/dfuse/pil4dfs/int_dfs.c:5609]: I'd probably make dcache_timeout be a uint32_t rather than uint16_t but don't call a function for one size and then change it, just use a 32 bit number throughout and set it to a sensible default when declared.
```suggestion
	rc = d_getenv_uint32_t("D_IL_DCACHE_TIMEOUT", &dcache_timeout);
	if (rc != -DER_NONEXIST)
	        ... handle fatal error.
```

**Review by ashleypittman — COMMENTED** [review]
There are lots of gaps here, the hash table will still grow forever except for the case of directories that are deleted locally, what about pruning of expired entries to speed lookup, what about entries that have been deleted by another process, what about entries which are longer supersets of ones that have been deleted.

The whole approach to using a hash table in this way is wrong, I don't think you should be making small adjustments to the implementation when this whole area will need to be rewritten anyway.

**Comment by mchaarawi** [comment]
> There are lots of gaps here, the hash table will still grow forever except for the case of directories that are deleted locally, what about pruning of expired entries to speed lookup, what about entries that have been deleted by another process, what about entries which are longer supersets of ones that have been deleted.
> 
> The whole approach to using a hash table in this way is wrong, I don't think you should be making small adjustments to the implementation when this whole area will need to be rewritten anyway.

I have not reviewed this yet to comment on the PR itself. it needs a rebase btw..
But i think the gaps are clear here in this hash implementation. the ultimate goal is to move the caching in the DFS layer where we can have a more proper implementation, but for now we need to address some of those gaps as we discussed several times in the WG. and it's better for now to address those in separate, smaller, and tested PRs, and document the ones that we don't get to for 2.6.

**Comment by ashleypittman** [comment]
> > There are lots of gaps here, the hash table will still grow forever except for the case of directories that are deleted locally, what about pruning of expired entries to speed lookup, what about entries that have been deleted by another process, what about entries which are longer supersets of ones that have been deleted.
> > The whole approach to using a hash table in this way is wrong, I don't think you should be making small adjustments to the implementation when this whole area will need to be rewritten anyway.
> 
> I have not reviewed this yet to comment on the PR itself. it needs a rebase btw.. But i think the gaps are clear here in this hash implementation. the ultimate goal is to move the caching in the DFS layer where we can have a more proper implementation, but for now we need to address some of those gaps as we discussed several times in the WG. and it's better for now to address those in separate, smaller, and tested PRs, and document the ones that we don't get to for 2.6.

I'm not 100% sure I understand your comment but let me rephrase what I meant above to see if that clarifies things.

The current hash table in pil4dfs contains paths, that is long strings with multiple intermediate directory names in it, so it might have "a", "a/b" and "a/b/c" all in the same hash table.  Adding logic to expire "a/b" from the hash table based on age does not move us forward, there is no point spending time improving anything around the hash table until it's been updated to store only dentries (parent + name pairs) and have pil4dfs doing the path walk properly for every lookup.  This code cannot exit it's technology preview state until that's resolved as an absolute minimum and this PR does nothing to move us closer to that point.

**Comment by ashleypittman** [comment]
> > The current hash table in pil4dfs contains paths, that is long strings with multiple intermediate directory names in it, so it might have "a", "a/b" and "a/b/c" all in the same hash table. Adding logic to expire "a/b" from the hash table based on age does not move us forward, there is no point spending time improving anything around the hash table until it's been updated to store only dentries (parent + name pairs) and have pil4dfs doing the path walk properly for every lookup. This code cannot exit it's technology preview state until that's resolved as an absolute minimum and this PR does nothing to move us closer to that point.
> 
> Thanks for the clarification. Yea I do agree with this part of your comment. I did not review the PR to see the age part of it and was just going by the title that deleted entries need to be removed from the cache. @wiliamhuang you do need to move away from this cache implementation that is used in IOR (use-case specific) and have finer grain lookups of each entry from the cache. As for the cache growing continuously, consistency with other processes, or background expiry of entries, this can be addressed in different PRs.

By "the cache growing continuously" what I mean is that with this PR in the example above "a/b" would only be removed from cache after in had expired during a lookup of "a/b" (which would presumably then put it back in the cache), there's no provision for removing expired entries that are not looked up although these would still be in the hash table which has an ongoing cost.

But - we need to correct what's in the hash table first before putting effort into optimising how the hash table is accessed or managed etc.

**Review by daltonbohning — COMMENTED** [review]
ftest code changes LGTM. Will defer on the C changes

**Comment by mchaarawi** [comment]
I believe this can be closed / deleted @wiliamhuang ?
